src/Processors/Formats/Impl/CSVRowInputFormat.cpp

```diff
@@ -36,6 +36,86 @@ CSVRowInputFormat::CSVRowInputFormat(
             ErrorCodes::BAD_ARGUMENTS);
 }
 
+void CSVRowInputFormat::readPrefix()
+{
+    if (with_names || with_types || data_types.at(0)->textCanContainOnlyValidUTF8())
+    {
+        /// We assume that column name or type cannot contain BOM, so, if format has header,
+        /// then BOM at beginning of stream cannot be confused with name or type of field, and it is safe to skip it.
+        skipBOMIfExists(*in);
+    }
+
+    /// This is a bit of abstraction leakage, but we need it in parallel parsing:
+    /// we check if this InputFormat is working with the "real" beginning of the data.
+    if (with_names && getCurrentUnitNumber() == 0)
+    {
+        if (format_settings.with_names_use_header)
+        {
+            std::vector<bool> read_columns(data_types.size(), false);
+
+            if (format_settings.csv.input_field_names.empty())
+            {
+                auto column_names = readNames();
+                for (const auto & name : column_names)
+                    addInputColumn(name, read_columns);
+            }
+            else
+            {
+                /// For Hive Text file, read the first row to get exact number of columns.
+                char * old_pos = in->position();
+                auto values = readHeaderRow();
+                in->position() = old_pos;
+
+                input_field_names = format_settings.csv.input_field_names;
+                input_field_names.resize(values.size());
+                for (const auto & column_name : input_field_names)
+                    addInputColumn(column_name, read_columns);
```

Comment:
This code won't work. While reading row we can call `next` in read buffer, we will lost all read data and won't be able to comeback to it (old_pos will point to new data in buffer or will be out of buffer at all). And why do we need to do resize input_field_names? input_field_names can contain only the first `k` column names? It sounds strange. But if you need such behaviour, you should use PeekableReadBuffer ro be able to comeback to some checkpoint

Reply:
Normaly `input_field_names` is the header of hive text file except one condition.

- Suppose hive table A has two columns b and c firstly. 
- Insert some data then file1 is generated in HDFS.
- Then we add a column named d in A(new column can only append to hive schema in hive)
- Insert some data then file2 is generated in HDFS.

In above condition, we get `input_field_names` from Hive Metastore, and it's `b, c, c` and  the header of file2 is `b, c, d`.
But the header of file1 is `b, c`, which is not the same with `input_field_names`, so we have to resize `inut_field_names`  when processing file1

Reply:
Bug fixed.